File: backend/app/services/compliance_scorer.py
import json
import re
import logging
from pathlib import Path
from app.config import get_settings
from app.models.schemas import (
    AccessibilityScore,
    ScoreDimension,
    CaptionBlock,
    VisualReferenceFlag,
)

logger = logging.getLogger(__name__)

VISUAL_REFERENCE_PATTERNS = [
    r"\bas you can see\b",
    r"\bas shown here\b",
    r"\blook at this\b",
    r"\bthis graph\b",
    r"\bthis chart\b",
    r"\bthis slide\b",
    r"\bthis diagram\b",
    r"\bthis table\b",
    r"\bthis image\b",
    r"\bover here\b",
    r"\bright here\b",
    r"\bon the screen\b",
    r"\bon the board\b",
    r"\bpoints to\b",
    r"\bhighlighted in\b",
    r"\bthe one in red\b",
    r"\bthe one in blue\b",
    r"\byou'll notice\b",
    r"\bif you look at\b",
]


class ComplianceScorer:
    """Score captions against the compliance rubric (WCAG 2.1 / Section 508 / DCMP)."""

    def __init__(self):
        self.settings = get_settings()
        rubric_path = Path(__file__).parent.parent.parent.parent / "docs" / "compliance-rubric.json"
        if rubric_path.exists():
            with open(rubric_path) as f:
                self.rubric = json.load(f)
        else:
            self.rubric = None

    def score_captions(
        self,
        captions: list[dict],
        duration_seconds: float | None = None,
    ) -> AccessibilityScore:
        dimensions = [
            self._score_accuracy(captions),
            self._score_synchronization(captions),
            self._score_completeness(captions, duration_seconds),
            self._score_speaker_identification(captions),
            self._score_formatting(captions),
            self._score_visual_accessibility(captions),
        ]

        overall = sum(d.score * d.weight for d in dimensions)

        if overall >= 95:
            rating = "Fully Compliant"
        elif overall >= 80:
            rating = "Mostly Compliant - Minor Issues"
        elif overall >= 60:
            rating = "Partially Compliant - Action Required"
        else:
            rating = "Non-Compliant - Significant Issues"

        return AccessibilityScore(
            overall=round(overall, 1),
            rating=rating,
            dimensions=dimensions,
        )

    def _score_accuracy(self, captions: list[dict]) -> ScoreDimension:
        issues = []
        total_words = 0
        problem_count = 0

        for cap in captions:
            text = cap.get("cleaned_text") or cap.get("original_text", "")
            words = text.split()
            total_words += len(words)

            if "[inaudible]" in text.lower():
                problem_count += text.lower().count("[inaudible]")
                issues.append(f"Caption {cap.get('sequence', '?')}: Contains [inaudible] segment")

            if re.search(r'\b[a-z]*[A-Z][a-z]*[A-Z]\b', text):
                problem_count += 1
                issues.append(f"Caption {cap.get('sequence', '?')}: Possible garbled text detected")

        score = max(0, 100 - (problem_count / max(total_words, 1)) * 10000)

        return ScoreDimension(
            id="accuracy",
            name="Caption Accuracy",
            score=min(100, round(score, 1)),
            weight=0.30,
            issues=issues[:20],
            details={"total_words": total_words, "problems_found": problem_count},
        )

    def _score_synchronization(self, captions: list[dict]) -> ScoreDimension:
        issues = []
        violations = 0
        total = len(captions)

        for cap in captions:
            start = cap.get("start_ms", 0)
            end = cap.get("end_ms", 0)
            duration = end - start

            if duration < self.settings.min_caption_duration_ms:
                violations += 1
                issues.append(
                    f"Caption {cap.get('sequence', '?')}: Duration {duration}ms "
                    f"below minimum {self.settings.min_caption_duration_ms}ms"
                )

            if duration > self.settings.max_caption_duration_ms:
                violations += 1
                issues.append(
                    f"Caption {cap.get('sequence', '?')}: Duration {duration}ms "
                    f"exceeds maximum {self.settings.max_caption_duration_ms}ms"
                )

            text = cap.get("cleaned_text") or cap.get("original_text", "")
            word_count = len(text.split())
            if duration > 0 and word_count > 0:
                wpm = (word_count / duration) * 60_000
                if wpm > self.settings.max_reading_speed_wpm:
                    violations += 1
                    issues.append(
                        f"Caption {cap.get('sequence', '?')}: Reading speed {wpm:.0f} WPM "
                        f"exceeds maximum {self.settings.max_reading_speed_wpm} WPM"
                    )

        score = max(0, 100 - (violations / max(total, 1)) * 100)

        return ScoreDimension(
            id="synchronization",
            name="Caption Synchronization",
            score=min(100, round(score, 1)),
            weight=0.20,
            issues=issues[:20],
            details={"total_captions": total, "timing_violations": violations},
        )

    def _score_completeness(
        self,
        captions: list[dict],
        duration_seconds: float | None = None,
    ) -> ScoreDimension:
        issues = []

        if not captions:
            return ScoreDimension(
                id="completeness",
                name="Caption Completeness",
                score=0,
                weight=0.20,
                issues=["No captions found"],
            )

        sorted_caps = sorted(captions, key=lambda c: c.get("start_ms", 0))
        gaps = []
        for i in range(1, len(sorted_caps)):
            prev_end = sorted_caps[i - 1].get("end_ms", 0)
            curr_start = sorted_caps[i].get("start_ms", 0)
            gap = curr_start - prev_end
            if gap > 5000:
                gaps.append((prev_end, curr_start, gap))
                issues.append(
                    f"Gap of {gap / 1000:.1f}s between captions "
                    f"{sorted_caps[i-1].get('sequence', '?')} and {sorted_caps[i].get('sequence', '?')}"
                )

        caption_coverage_ms = sum(
            max(0, c.get("end_ms", 0) - c.get("start_ms", 0)) for c in captions
        )

        if duration_seconds and duration_seconds > 0:
            coverage_pct = (caption_coverage_ms / (duration_seconds * 1000)) * 100
        else:
            last_end = max(c.get("end_ms", 0) for c in captions) if captions else 0
            coverage_pct = (caption_coverage_ms / max(last_end, 1)) * 100

        gap_penalty = len(gaps) * 5
        score = max(0, min(100, coverage_pct) - gap_penalty)

        return ScoreDimension(
            id="completeness",
            name="Caption Completeness",
            score=round(score, 1),
            weight=0.20,
            issues=issues[:20],
            details={
                "coverage_pct": round(coverage_pct, 1),
                "large_gaps": len(gaps),
                "caption_count": len(captions),
            },
        )

    def _score_speaker_identification(self, captions: list[dict]) -> ScoreDimension:
        issues = []

        speakers = set()
        labeled_count = 0
        for cap in captions:
            if cap.get("speaker"):
                speakers.add(cap["speaker"])
                labeled_count += 1

        total = len(captions)
        if total == 0:
            return ScoreDimension(
                id="speaker_identification",
                name="Speaker Identification",
                score=0,
                weight=0.10,
                issues=["No captions to evaluate"],
            )

        if len(speakers) <= 1 and labeled_count == 0:
            score = 80.0
            issues.append(
                "No speaker labels detected. If this is a single-speaker lecture, "
                "this is acceptable. If multiple speakers are present, labels are required."
            )
        elif len(speakers) > 1:
            label_ratio = labeled_count / total
            score = label_ratio * 100
            if label_ratio < 1.0:
                issues.append(
                    f"Only {labeled_count}/{total} captions have speaker labels. "
                    f"All captions should have speaker identification when multiple speakers are present."
                )
        else:
            score = 100.0

        return ScoreDimension(
            id="speaker_identification",
            name="Speaker Identification",
            score=round(score, 1),
            weight=0.10,
            issues=issues,
            details={"speakers_found": len(speakers), "labeled_captions": labeled_count},
        )

    def _score_formatting(self, captions: list[dict]) -> ScoreDimension:
        issues = []
        violations = 0
        total = len(captions)

        for cap in captions:
            text = cap.get("cleaned_text") or cap.get("original_text", "")
            lines = text.split("\n")

            if len(lines) > self.settings.max_caption_lines:
                violations += 1
                issues.append(
                    f"Caption {cap.get('sequence', '?')}: {len(lines)} lines "
                    f"(max {self.settings.max_caption_lines})"
                )

            for line in lines:
                if len(line) > self.settings.max_caption_line_length:
                    violations += 1
                    issues.append(
                        f"Caption {cap.get('sequence', '?')}: Line has {len(line)} chars "
                        f"(max {self.settings.max_caption_line_length})"
                    )
                    break

            if text == text.upper() and len(text) > 10:
                violations += 1
                issues.append(f"Caption {cap.get('sequence', '?')}: ALL CAPS text (should be mixed case)")

        score = max(0, 100 - (violations / max(total, 1)) * 100)

        return ScoreDimension(
            id="formatting",
            name="Caption Formatting",
            score=min(100, round(score, 1)),
            weight=0.10,
            issues=issues[:20],
            details={"total_captions": total, "formatting_violations": violations},
        )

    def _score_visual_accessibility(self, captions: list[dict]) -> ScoreDimension:
        issues = []
        flags: list[VisualReferenceFlag] = []

        for cap in captions:
            text = cap.get("cleaned_text") or cap.get("original_text", "")
            for pattern in VISUAL_REFERENCE_PATTERNS:
                matches = re.finditer(pattern, text, re.IGNORECASE)
                for match in matches:
                    flags.append(VisualReferenceFlag(
                        caption_id=cap.get("id", ""),
                        text=text,
                        matched_pattern=match.group(),
                        start_ms=cap.get("start_ms", 0),
                        suggestion=f"Add a description of the visual content referenced by '{match.group()}'",
                    ))
                    issues.append(
                        f"Caption {cap.get('sequence', '?')}: Visual reference '{match.group()}' "
                        f"needs description for accessibility"
                    )

        if flags:
            score = max(0, 100 - len(flags) * 10)
        else:
            score = 100.0

        return ScoreDimension(
            id="visual_accessibility",
            name="Visual Reference Accessibility",
            score=min(100, round(score, 1)),
            weight=0.10,
            issues=issues[:20],
            details={"visual_references_found": len(flags)},
        )

    def detect_all_visual_references(self, captions: list[dict]) -> list[VisualReferenceFlag]:
        """Return all detected visual references for the UI to display."""
        flags = []
        for cap in captions:
            text = cap.get("cleaned_text") or cap.get("original_text", "")
            for pattern in VISUAL_REFERENCE_PATTERNS:
                for match in re.finditer(pattern, text, re.IGNORECASE):
                    flags.append(VisualReferenceFlag(
                        caption_id=cap.get("id", ""),
                        text=text,
                        matched_pattern=match.group(),
                        start_ms=cap.get("start_ms", 0),
                        suggestion=f"Add a description of the visual content referenced by '{match.group()}'",
                    ))
        return flags

File: frontend/src/lib/server/compliance-scorer.ts
import { config } from "./config";

export interface ScoreDimension {
  id: string;
  name: string;
  score: number;
  weight: number;
  issues: string[];
  details: Record<string, unknown>;
}

export interface AccessibilityScore {
  overall: number;
  rating: string;
  dimensions: ScoreDimension[];
}

export interface VisualReferenceFlag {
  caption_id: string;
  text: string;
  matched_pattern: string;
  start_ms: number;
  suggestion: string;
}

const VISUAL_REFERENCE_PATTERNS = [
  /\bas you can see\b/i,
  /\bas shown here\b/i,
  /\blook at this\b/i,
  /\bthis graph\b/i,
  /\bthis chart\b/i,
  /\bthis slide\b/i,
  /\bthis diagram\b/i,
  /\bthis table\b/i,
  /\bthis image\b/i,
  /\bover here\b/i,
  /\bright here\b/i,
  /\bon the screen\b/i,
  /\bon the board\b/i,
  /\bpoints to\b/i,
  /\bhighlighted in\b/i,
  /\bthe one in red\b/i,
  /\bthe one in blue\b/i,
  /\byou'll notice\b/i,
  /\bif you look at\b/i,
];

type Cap = Record<string, unknown>;

function getText(cap: Cap): string {
  return ((cap.cleaned_text as string) || (cap.original_text as string) || "");
}

export function scoreCaptions(
  captions: Cap[],
  durationSeconds?: number | null
): AccessibilityScore {
  const dimensions = [
    scoreAccuracy(captions),
    scoreSynchronization(captions),
    scoreCompleteness(captions, durationSeconds),
    scoreSpeakerIdentification(captions),
    scoreFormatting(captions),
    scoreVisualAccessibility(captions),
  ];

  const overall = dimensions.reduce((s, d) => s + d.score * d.weight, 0);

  let rating: string;
  if (overall >= 95) rating = "Fully Compliant";
  else if (overall >= 80) rating = "Mostly Compliant - Minor Issues";
  else if (overall >= 60) rating = "Partially Compliant - Action Required";
  else rating = "Non-Compliant - Significant Issues";

  return { overall: Math.round(overall * 10) / 10, rating, dimensions };
}

function scoreAccuracy(captions: Cap[]): ScoreDimension {
  const issues: string[] = [];
  let totalWords = 0;
  let problemCount = 0;

  for (const cap of captions) {
    const text = getText(cap);
    const words = text.split(/\s+/);
    totalWords += words.length;

    if (text.toLowerCase().includes("[inaudible]")) {
      problemCount += (text.toLowerCase().match(/\[inaudible\]/g) ?? []).length;
      issues.push(`Caption ${cap.sequence ?? "?"}: Contains [inaudible] segment`);
    }
    if (/\b[a-z]*[A-Z][a-z]*[A-Z]\b/.test(text)) {
      problemCount++;
      issues.push(`Caption ${cap.sequence ?? "?"}: Possible garbled text detected`);
    }
  }

  const score = Math.max(0, 100 - (problemCount / Math.max(totalWords, 1)) * 10000);
  return {
    id: "accuracy",
    name: "Caption Accuracy",
    score: Math.min(100, Math.round(score * 10) / 10),
    weight: 0.30,
    issues: issues.slice(0, 20),
    details: { total_words: totalWords, problems_found: problemCount },
  };
}

function scoreSynchronization(captions: Cap[]): ScoreDimension {
  const issues: string[] = [];
  let violations = 0;
  const total = captions.length;

  for (const cap of captions) {
    const start = (cap.start_ms as number) ?? 0;
    const end = (cap.end_ms as number) ?? 0;
    const duration = end - start;

    if (duration < config.minCaptionDurationMs) {
      violations++;
      issues.push(`Caption ${cap.sequence ?? "?"}: Duration ${duration}ms below minimum ${config.minCaptionDurationMs}ms`);
    }
    if (duration > config.maxCaptionDurationMs) {
      violations++;
      issues.push(`Caption ${cap.sequence ?? "?"}: Duration ${duration}ms exceeds maximum ${config.maxCaptionDurationMs}ms`);
    }

    const text = getText(cap);
    const wordCount = text.split(/\s+/).length;
    if (duration > 0 && wordCount > 0) {
      const wpm = (wordCount / duration) * 60_000;
      if (wpm > config.maxReadingSpeedWpm) {
        violations++;
        issues.push(`Caption ${cap.sequence ?? "?"}: Reading speed ${Math.round(wpm)} WPM exceeds maximum ${config.maxReadingSpeedWpm} WPM`);
      }
    }
  }

  const score = Math.max(0, 100 - (violations / Math.max(total, 1)) * 100);
  return {
    id: "synchronization",
    name: "Caption Synchronization",
    score: Math.min(100, Math.round(score * 10) / 10),
    weight: 0.20,
    issues: issues.slice(0, 20),
    details: { total_captions: total, timing_violations: violations },
  };
}

function scoreCompleteness(captions: Cap[], durationSeconds?: number | null): ScoreDimension {
  if (!captions.length) {
    return { id: "completeness", name: "Caption Completeness", score: 0, weight: 0.20, issues: ["No captions found"], details: {} };
  }

  const issues: string[] = [];
  const sorted = [...captions].sort((a, b) => ((a.start_ms as number) ?? 0) - ((b.start_ms as number) ?? 0));
  let gaps = 0;

  for (let i = 1; i < sorted.length; i++) {
    const prevEnd = (sorted[i - 1].end_ms as number) ?? 0;
    const currStart = (sorted[i].start_ms as number) ?? 0;
    const gap = currStart - prevEnd;
    if (gap > 5000) {
      gaps++;
      issues.push(`Gap of ${(gap / 1000).toFixed(1)}s between captions ${sorted[i - 1].sequence ?? "?"} and ${sorted[i].sequence ?? "?"}`);
    }
  }

  const captionCoverageMs = captions.reduce(
    (s, c) => s + Math.max(0, ((c.end_ms as number) ?? 0) - ((c.start_ms as number) ?? 0)),
    0
  );

  let coveragePct: number;
  if (durationSeconds && durationSeconds > 0) {
    coveragePct = (captionCoverageMs / (durationSeconds * 1000)) * 100;
  } else {
    const lastEnd = Math.max(...captions.map((c) => (c.end_ms as number) ?? 0));
    coveragePct = (captionCoverageMs / Math.max(lastEnd, 1)) * 100;
  }

  const score = Math.max(0, Math.min(100, coveragePct) - gaps * 5);
  return {
    id: "completeness",
    name: "Caption Completeness",
    score: Math.round(score * 10) / 10,
    weight: 0.20,
    issues: issues.slice(0, 20),
    details: { coverage_pct: Math.round(coveragePct * 10) / 10, large_gaps: gaps, caption_count: captions.length },
  };
}

function scoreSpeakerIdentification(captions: Cap[]): ScoreDimension {
  const issues: string[] = [];
  const speakers = new Set<string>();
  let labeledCount = 0;

  for (const cap of captions) {
    if (cap.speaker) {
      speakers.add(cap.speaker as string);
      labeledCount++;
    }
  }

  const total = captions.length;
  if (!total) {
    return { id: "speaker_identification", name: "Speaker Identification", score: 0, weight: 0.10, issues: ["No captions to evaluate"], details: {} };
  }

  let score: number;
  if (speakers.size <= 1 && labeledCount === 0) {
    score = 80;
    issues.push("No speaker labels detected. If this is a single-speaker lecture, this is acceptable.");
  } else if (speakers.size > 1) {
    const ratio = labeledCount / total;
    score = ratio * 100;
    if (ratio < 1) {
      issues.push(`Only ${labeledCount}/${total} captions have speaker labels.`);
    }
  } else {
    score = 100;
  }

  return {
    id: "speaker_identification",
    name: "Speaker Identification",
    score: Math.round(score * 10) / 10,
    weight: 0.10,
    issues,
    details: { speakers_found: speakers.size, labeled_captions: labeledCount },
  };
}

function scoreFormatting(captions: Cap[]): ScoreDimension {
  const issues: string[] = [];
  let violations = 0;
  const total = captions.length;

  for (const cap of captions) {
    const text = getText(cap);
    const lines = text.split("\n");

    if (lines.length > config.maxCaptionLines) {
      violations++;
      issues.push(`Caption ${cap.sequence ?? "?"}: ${lines.length} lines (max ${config.maxCaptionLines})`);
    }
    for (const line of lines) {
      if (line.length > config.maxCaptionLineLength) {
        violations++;
        issues.push(`Caption ${cap.sequence ?? "?"}: Line has ${line.length} chars (max ${config.maxCaptionLineLength})`);
        break;
      }
    }
    if (text === text.toUpperCase() && text.length > 10) {
      violations++;
      issues.push(`Caption ${cap.sequence ?? "?"}: ALL CAPS text (should be mixed case)`);
    }
  }

  const score = Math.max(0, 100 - (violations / Math.max(total, 1)) * 100);
  return {
    id: "formatting",
    name: "Caption Formatting",
    score: Math.min(100, Math.round(score * 10) / 10),
    weight: 0.10,
    issues: issues.slice(0, 20),
    details: { total_captions: total, formatting_violations: violations },
  };
}

function scoreVisualAccessibility(captions: Cap[]): ScoreDimension {
  const issues: string[] = [];
  let flagCount = 0;

  for (const cap of captions) {
    const text = getText(cap);
    for (const pattern of VISUAL_REFERENCE_PATTERNS) {
      const matches = text.matchAll(new RegExp(pattern.source, "gi"));
      for (const match of matches) {
        flagCount++;
        issues.push(`Caption ${cap.sequence ?? "?"}: Visual reference '${match[0]}' needs description for accessibility`);
      }
    }
  }

  const score = flagCount ? Math.max(0, 100 - flagCount * 10) : 100;
  return {
    id: "visual_accessibility",
    name: "Visual Reference Accessibility",
    score: Math.min(100, Math.round(score * 10) / 10),
    weight: 0.10,
    issues: issues.slice(0, 20),
    details: { visual_references_found: flagCount },
  };
}

export function detectAllVisualReferences(captions: Cap[]): VisualReferenceFlag[] {
  const flags: VisualReferenceFlag[] = [];
  for (const cap of captions) {
    const text = getText(cap);
    for (const pattern of VISUAL_REFERENCE_PATTERNS) {
      for (const match of text.matchAll(new RegExp(pattern.source, "gi"))) {
        flags.push({
          caption_id: (cap.id as string) ?? "",
          text,
          matched_pattern: match[0],
          start_ms: (cap.start_ms as number) ?? 0,
          suggestion: `Add a description of the visual content referenced by '${match[0]}'`,
        });
      }
    }
  }
  return flags;
}
